plugins/soundsourcem4a/soundsourcem4a.cpp

```diff
@@ -372,107 +371,100 @@ void SoundSourceM4A::restartDecoding(MP4SampleId sampleBlockId) {
+            if ((kSampleBlockIdMin + m_numberOfPrefetchSampleBlocks)
+                    < sampleBlockId) {
+                sampleBlockId -= m_numberOfPrefetchSampleBlocks;
+            } else {
+                sampleBlockId = kSampleBlockIdMin;
+            }
+            restartDecoding(sampleBlockId);
+            DEBUG_ASSERT(m_curSampleBlockId == sampleBlockId);
+        }
 
-    // Skip (= decode and discard) all samples up to the target position
-    const SINT prefetchFrameCount = frameIndex - m_curFrameIndex;
-    const SINT skipFrameCount = skipSampleFrames(prefetchFrameCount);
-    DEBUG_ASSERT(skipFrameCount <= prefetchFrameCount);
-    if (skipFrameCount < prefetchFrameCount) {
-        kLogger.warning() << "Failed to prefetch sample data while seeking"
-                << skipFrameCount << "<" << prefetchFrameCount;
+        // Decoding starts before the actual target position
+        DEBUG_ASSERT(m_curFrameIndex <= firstFrameIndex);
+        const auto precedingFrames =
+                IndexRange::between(m_curFrameIndex, firstFrameIndex);
+        if (!precedingFrames.empty()
+                && (precedingFrames != readSampleFramesClamped(
+                        WritableSampleFrames(precedingFrames)).frameIndexRange())) {
+            kLogger.warning()
+                    << "Failed to skip preceding frames"
+                    << precedingFrames;
+            // Abort
+            return ReadableSampleFrames(
+                    IndexRange::between(
+                            m_curFrameIndex,
+                            m_curFrameIndex));
+        }
     }
+    DEBUG_ASSERT(m_curFrameIndex == firstFrameIndex);
 
-    DEBUG_ASSERT(isValidFrameIndex(m_curFrameIndex));
-    return m_curFrameIndex;
-}
-
-SINT SoundSourceM4A::readSampleFrames(
-        SINT numberOfFrames, CSAMPLE* sampleBuffer) {
-    DEBUG_ASSERT(isValidFrameIndex(m_curFrameIndex));
-
-    const SINT numberOfFramesTotal = math_min(
-            numberOfFrames, getMaxFrameIndex() - m_curFrameIndex);
-    const SINT numberOfSamplesTotal = frames2samples(numberOfFramesTotal);
+    const SINT numberOfSamplesTotal = frames2samples(writableSampleFrames.frameIndexRange().length());
 
-    CSAMPLE* pSampleBuffer = sampleBuffer;
     SINT numberOfSamplesRemaining = numberOfSamplesTotal;
+    SINT outputSampleOffset = 0;
     while (0 < numberOfSamplesRemaining) {
 
-        if (!m_sampleBuffer.isEmpty()) {
+        if (!m_sampleBuffer.empty()) {
             // Consume previously decoded sample data
-            const SampleBuffer::ReadableChunk readableChunk(
-                    m_sampleBuffer.readFromHead(numberOfSamplesRemaining));
-            if (pSampleBuffer) {
-                SampleUtil::copy(pSampleBuffer, readableChunk.data(), readableChunk.size());
-                pSampleBuffer += readableChunk.size();
+            const SampleBuffer::ReadableSlice readableSlice(
+                    m_sampleBuffer.readLifo(numberOfSamplesRemaining));
```

Comment:
We have a single buffer here, but read Lifo implies there is an underlying Lifo. So it looks like we swap samples, because the last samples we put in where read first? 
readFromHead() was better, but does is actually swap things?  
Is there disadvantage to write: 
const SampleBuffer::ReadableSlice readableSlice = m_sampleBuffer.readLifo(numberOfSamplesRemaining);